### PR TITLE
Drbg rework (part 1)

### PR DESCRIPTION
**Pull request by onvej-sl**
This PR replaces https://github.com/trezor/trezor-firmware/pull/1481.

The purpose of this PR is to replace hmac_drbg (that is used by random interrupts and `wait_random`) with faster chacha_drbg.

### CODE REVIEW

**Comment by onvej-sl** [comment]
These are comments unresolved in the original PR that are hopefully resolved in this PR:
https://github.com/trezor/trezor-firmware/pull/1481#discussion_r578536056
https://github.com/trezor/trezor-firmware/pull/1481#discussion_r586517905
https://github.com/trezor/trezor-firmware/pull/1481#discussion_r593026121

**Review by andrewkozlik — CHANGES_REQUESTED** [review]
It still doesn't make sense to me why the original pull request was closed after having been extensively reviewed. I had to redo this entire review from scratch and there is more to come. Terribly inefficient.

File: legacy/random_delays.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef __RANDOM_DELYAS_H__
+#define __RANDOM_DELYAS_H__
```
[the remark this answers is not in the document]
Comment (onvej-sl) [reply legacy/random_delays.h]: 20f91a71ac61329a8229ad44cd3c79382340df9e

File: core/embed/trezorhal/random_delays.c [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    drbg_reseed();
+  }
+  chacha_drbg_generate(&drbg_ctx, buffer, length);
+}
+
+static uint32_t drbg_random8(void) {
+  // Since the function is called both from an interrupt (rdi_handler,
+  // wait_random) and the main thread (wait_random), we use a lock to
+  // synchronise access to global variables
+  static volatile atomic_flag locked = {0};
```
[the remark this answers is not in the document]
Comment (onvej-sl) [reply core/embed/trezorhal/random_delays.c]: d1164f485a291227bb00dd184f67a7f14ca4bcc5

File: core/embed/trezorhal/random_delays.c [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+static void drbg_generate(uint8_t *buffer, size_t length) {
+  ensure(drbg_initialized, NULL);
+
+  if (drbg_ctx.reseed_counter > DRBG_RESEED_INTERVAL_CALLS) {
+    drbg_reseed();
+  }
+  chacha_drbg_generate(&drbg_ctx, buffer, length);
+}
+
+static uint32_t drbg_random8(void) {
```
[the remark this answers is not in the document]
Comment (onvej-sl) [reply core/embed/trezorhal/random_delays.c:90]: d1164f485a291227bb00dd184f67a7f14ca4bcc5

File: core/embed/trezorhal/random_delays.c [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+      "add r0, $1;"  // wait one tick
+      "wait_0:"
+      :
+      : "m"(delay)
+      : "r0", "r1");
+}
+
+void random_delays_init() { drbg_init(); }
+
+void rdi_start(void) {
```
[the remark this answers is not in the document]
Comment (onvej-sl) [reply core/embed/trezorhal/random_delays.c:155]: Definitely.

Done in d1164f485a291227bb00dd184f67a7f14ca4bcc5.

**Review by andrewkozlik — APPROVED** [review]

**Pull request merged**